lib_eio_js/browser/eio_browser.ml

```diff
@@ -0,0 +1,198 @@
+
+  let push_head q v = ignore (Lwt_dllist.add_l v q : 'a Lwt_dllist.node)
+
+  let pop q = Lwt_dllist.take_opt_l q
+end
+
+module Ctf = Eio.Private.Ctf
+
+module Suspended = struct
+  type 'a t = {
+    fiber : Eio.Private.Fiber_context.t;
+    k : ('a, unit) Effect.Deep.continuation;
+  }
+
+  let tid t = Eio.Private.Fiber_context.tid t.fiber
+
+  let continue t v =
+    Ctf.note_switch (tid t);
+    Effect.Deep.continue t.k v
+
+  let discontinue t ex =
+    Ctf.note_switch (tid t);
+    Effect.Deep.discontinue t.k ex
+end
+
+module Scheduler = struct
+  type t = {
+    listener : Brr.Ev.listener;
+    schedule : (unit -> unit);
+  }
+
+  let v ~schedule t =
+    let open Brr_io in
+    let scheduler = El.div [] in
+    let el = El.as_target scheduler in
+    let listener = Brr.Ev.listen Message.Ev.message (fun _ev -> schedule t) el in
+    let schedule =
+      let args = [| Ev.create Message.Ev.message |> Ev.to_jv |] in
+      let sched = El.to_jv scheduler in
+      fun () -> Jv.call sched "dispatchEvent" args |> ignore
+    in
+    { listener; schedule }
+
+  let stop t =
+    Brr.Ev.unlisten t.listener
+
+  let wakeup t = t.schedule ()
+end
+
+type t = {
+  (* Suspended fibers waiting to run again.
+     [Run_queue] is like [Stdlib.Queue], but allows pushing items
+     to the head too, which we need. *)
+  mutable run_q : (unit -> unit) Run_queue.t;
+  mutable pending_io : int;
+  mutable scheduler : Scheduler.t option;
+}
+
+let enqueue_thread t k v =
+  Run_queue.push t.run_q (fun () -> Suspended.continue k v);
+  t.pending_io <- t.pending_io - 1;
+  Option.iter Scheduler.wakeup t.scheduler
+
+let enqueue_failed_thread t k v =
+  Run_queue.push t.run_q (fun () -> Suspended.discontinue k v);
+  t.pending_io <- t.pending_io - 1;
+  Option.iter Scheduler.wakeup t.scheduler
+
+
+type _ Effect.t += Enter_unchecked : (t -> 'a Suspended.t -> unit) -> 'a Effect.t
+let enter_unchecked fn = Effect.perform (Enter_unchecked fn)
+
+let enter_io fn =
+  enter_unchecked @@ fun st k ->
+  st.pending_io <- st.pending_io + 1;
+  fn st k
+
+(* Resume the next runnable fiber, if any. *)
+let schedule t : unit =
+  match Run_queue.pop t.run_q with
+  | Some f -> f ();
+  | None ->
+    if t.pending_io = 0 then begin
+      Option.iter Scheduler.stop t.scheduler
+    end
+
+module Timeout = struct
+  let sleep ~ms =
+    enter_io @@ fun st k ->
+    let timer = ref None in
+    Fiber_context.set_cancel_fn k.fiber (fun exn -> Option.iter G.stop_timer !timer; enqueue_failed_thread st k exn);
+    let id = G.set_timeout ~ms (fun () ->
+        Fiber_context.clear_cancel_fn k.fiber;
+        enqueue_thread st k ()
+      ) in
+    timer := Some id
+end
+
+let await fut =
+  enter_io @@ fun st k ->
+  let cancelled = ref true in
+  Fiber_context.set_cancel_fn k.fiber (fun exn -> cancelled := true; enqueue_failed_thread st k exn);
+  Fut.await fut (fun v ->
+      Fiber_context.clear_cancel_fn k.fiber;
```

Comment:
This looks wrong. If I understand correctly, cancellation will go like this:
1. Someone creates a `Fut` and `await`s it.
2. The await is cancelled. `cancelled` is set to `true` and the waiter is resumed.
3. The fiber starts a new operation.
4. `Fut` is later resolved. It (incorrectly) clears the cancel function of the fiber's new operation.

Fixing that just requires putting the clear inside the `if`.

It's also a bit of a problem that we leak memory every time we await and then cancel (until the Fut is finally resolved). Eio generally tries to avoid doing that, but it may be impossible with the Fut API.

Reply:
Yep quite right! I've added a comment about the memory leak. I'm not sure there's a way around this because `Fut` is based on JS promises and they are eager so don't have a way to cancel them iiuc.